Sprint 07

aldadametov opened the pull request:
1. Написаны unit-тесты:
-На сабскрипт для безопасного получения элемента массива по индексу (ArrayTests);
-На сервис MoviesLoader (для проверки успешной загрузки и для проверки ошибки MoviesLoaderTests).
2. Написаны и проходят UI-тесты:
-Два теста на замену одного постера на другой при нажатии кнопки ответа — 
Тест кнопки Да (при нажатии на кнопку Да меняется лейбл с номером вопроса).
Тест кнопки Нет (при нажатии на кнопку Нет меняется лейбл с номером вопроса).
-Тест появления алерта при окончании раунда (вам нужно проверить, что алерт появился, протестировать текст на кнопке и в заголовке алерта).
Т-ест скрытия алерта после нажатия на кнопку на нём (в этом тесте нужно проверить, что алерт отсутствует на экране и счётчик вопросов сбросился к значению 1/10).
3. Проведён рефакторинг проекта — отрефакторен MovieQuizViewController на архитектуру MVP 
4. Написан тест для convert(model: QuizQuestion) -> QuizStepViewModel.

aldadametov committed:
1. Добавлены Unit тесты для Array
2. Добавлены UI тесты для MoviesLoader

aldadametov committed:
1. Произведена первая часть рефакторинга, часть логики перенесена из MovieQuizViewController в MovieQuizViewPresenter

aldadametov committed:
Логика обработки нажатия на yesButton и noButton вынесена в Presenter

aldadametov committed:
метод showNextQuestionOrResults вынесен в MovieQuizPresenter

aldadametov committed:
MovieQuizPresenter теперь является делегатом для QuestionFactory

aldadametov committed:
Проведён рефакторинг проекта — отрефакторен MovieQuizViewController на архитектуру MVP
Написан и проходит UNIT тест для convert(model: QuizQuestion) -> QuizStepViewModel.

IBDesignable commented on MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -8,9 +8,9 @@ import UIKit
+    func show(from viewController: UIViewController, with model: AlertModel) {
         let alert = UIAlertController(title: model.title, message: model.message, preferredStyle: .alert)
-        
+        alert.view.accessibilityIdentifier = "Game Results"
```
👍🏻

IBDesignable commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,140 @@
+//  Created by Алишер Дадаметов on 17.05.2023.
+//
+
+import UIKit
```
В любой архитекуре (MVP, VIPER, MVVM), обязательно есть слой логики, в котором UIKit проникать не должен. В данном случае это Presenter. Давай удалим, посмотрим можно ли без импорта собраться, и при необходимости доправим код таким образом, чтобы UI-китовые сущности сюда не проникали.

Про это помни и в следующих спринтах при работе с архитектурой

IBDesignable commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,140 @@
+    
+    // MARK: - QuestionFactoryDelegate
+        
+    func didLoadDataFromServer() {
+        viewController?.hideLoadingIndicator()
+        questionFactory?.requestNextQuestion()
+        }
+        
```
Поправить отступы можно так: выедли код и зажми control+i

IBDesignable commented on MovieQuizUITests/MovieQuizUITests.swift:
```diff
@@ -0,0 +1,111 @@
+    
+    func testGameFinish() {
+        sleep(2)
+        for _ in 1...10 {
+            app.buttons["No"].tap()
+            sleep(2)
+        }
+
```
этот участок в текстах повторяется два раза. В тестах тоже можно писать приватные функции, чтобы выделять обшую часть.

IBDesignable commented on MovieQuizUITests/MovieQuizUITests.swift:
```diff
@@ -0,0 +1,111 @@
+        
+        
+    }
+    
+        
+    
+    func testExample() throws {
+        let app = XCUIApplication()
+        app.launch()
+    }
```
этот тест не несет никакой смысловой нагрузки: надо удалить

aldadametov committed:
Убраны лишние отступы и пропуски строк
Теперь индикатор загрузки показывается и между вопросами

aldadametov committed:
1. Статус бар теперь белого цвета
2, Пользователь не может прожать кнопку больше одного раза за один показ вопроса
